src/clear_display.c

```diff
@@ -48,20 +48,67 @@
 #include "tlf_curses.h"
 #include "ui_utils.h"
 
-static char terminal1[88] = "";
-static char terminal2[88] = "";
-static char terminal3[88] = "";
-static char terminal4[88] = "";
+void clear_display(void);
 
 
-void init_terminal_strings(void) {
-    strcat(terminal1, backgrnd_str);
-    strcat(terminal2, backgrnd_str);
-    strcat(terminal3, backgrnd_str);
-    strcat(terminal4, backgrnd_str);
+void clear_line(int row) {
+    mvaddstr(row, 0, backgrnd_str);
+}
+
+static char *terminal1;
+static char *terminal2;
+static char *terminal3;
+static char *terminal4;
+
+void init_keyer_terminal(void) {
+    terminal1 = g_strdup("");
+    terminal2 = g_strdup("");
+    terminal3 = g_strdup("");
+    terminal4 = g_strdup("");
+}
+
+static void show_keyer_terminal() {
+    attron(modify_attr(COLOR_PAIR(C_LOG) | A_STANDOUT));
+    for (int i = 1; i < 6; i++)
+	clear_line(i);
+
+    mvaddstr(1, 0, terminal1);
+    mvaddstr(2, 0, terminal2);
+    mvaddstr(3, 0, terminal3);
+    mvaddstr(4, 0, terminal4);
+}
+
+/*
+ * add 'buffer' to the keyer terminal while scrolling one line up
+ * and refreshing the display
+ */
+static void add_line_to_keyer_terminal(char *buffer) {
+
+    g_free(terminal1);
+    terminal1 = terminal2;
+    terminal2 = terminal3;
+    terminal3 = terminal4;
+
+    terminal4 = g_strdup(buffer);
+    g_strchomp(terminal4);
+
+    move(5, 0);
+
+    clear_display();
+}
+
+/*
+ * add 'buffer' to keyer terminal, splitting it into separate lines if needed
+ */
+void add_to_keyer_terminal(char *buffer) {
+    gchar **lines = g_strsplit_set(buffer, "\n\r", 0);
+    for (int i = 0; lines[i] != NULL; i++)
+        add_line_to_keyer_terminal(lines[i]);
```

Comment:
Using this setting
```
F1= CQ CQ DE % % %|CQ CQ DX % % % K
```
and calling CQ twice Tlf shows this
![image](https://user-images.githubusercontent.com/15141948/147282839-6871f9b4-cd6c-4f35-83da-ae024e3f428b.png)

whereas fldigi shows
![image](https://user-images.githubusercontent.com/15141948/147282879-c1ce48f7-d55e-4376-9b04-42de143e4d7f.png)

There seem to be a leading CR in the digi keyer strings which leads to an empty line on Tlf keyer terminal.
How about moving the chomp'ing into the for loop before calling `add_line_to_keyer_terminal` and skipping if line is empty? This of course means not displaying embedded CRs ("text1|||text2"), but we have just 4 lines anyway...

Reply:
Good point. I agree - the extra newline has to be dropped. Question is if we drop embedded newlines / carriage returns completely or cut them down to at most one newline.

Will look into it tomorrow. Today are other things to do :-) here.

Reply:
Actual solution is as recommended - just drop empty lines from display. They will still be sent via the digit keyer.

As far as I remember the discussion some years ago digi mode users are very seldom use more newlines than really needed. They try to use vertical space very sparingly - maybe leftover from paper printers :-).

Reply:
Thanks, works OK.
HNY!